### PR TITLE
Feature/upload component

### LINKED ISSUE #1533: Create and style assignment upload component
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### PR DESCRIPTION
**Pull request by stevenbister**
### 🧹 Please check off the following before submitting:

- [x] Close issues
- [x] Request reviewers
- [x] Assign yourself
- [ ] Add labels (e.g. `build:osc-ecommerce`)
- [x] Add project
- [x] Set project status
- [x] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #1533 

## 📝 Description

Adds the file input component.

## ⛳️ Current behaviour (updates)

N/a

## 🚀 New behaviour

Adds file input component, with stories and tests

## 💣 Is this a breaking change (Yes/No): No

> Please describe any breaking changes.

## 🧾 Final Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible cross-device impacts (iOS, Android)
- [x] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [x] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.

## 📝 Additional Information

There's one test in there that's commented out. I cannot for the life of me get it to work 😅 But I've left it in, if anyone has any ideas feel free to give them a go. Otherwise I might have another pop at it later down the line.

### Summary by CodeRabbit

- New Feature: Introduced a new `FileInput` component with drag-and-drop functionality, client-side validation, and error handling. This component supports various properties such as `id`, `icon`, `label`, `name`, `multiple`, `accept`, etc.
- New Feature: Added a `FileInputError` component to display error messages associated with the file input.
- Test: Added comprehensive test cases for the `FileInput` component, covering rendering, validation, error messages, and drag events.
- Chore: Updated the `clientSideValidation` function to support `FileList` type inputs.
- Documentation: Added Storybook stories for the `FileInput` component showcasing different usage scenarios.

### CODE REVIEW

File: packages/osc-ui/src/components/FileInput/FileInput.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
+import { createEvent, fireEvent, screen } from '@testing-library/react';
+import React from 'react';
+import { render } from 'test-utils';
+import { FileInput } from './FileInput';
+import { fileInputSchema } from './mockSchema';
+
+afterEach(() => {
+    vi.clearAllMocks();
+});
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-ui/src/components/FileInput/FileInput.spec.tsx:9]: This is incorrect we're using vitest not jest

**Review by mattwendzina — COMMENTED** [review]
Looks great mate. A couple of comments/questions in regards to the functionality.

- Looks like you can't remove an upload once you've added it? Think this could be quite important/helpful as it's easy to accidentally at the wrong file or something.
- Is there a reason there is a pop up once it's been uploaded? Doesn't feel necessary to me given that you see the name of the file come up up in the download box anyway. 
- Doesn't look like you can upload multiple files at the same time - If I add a second it just overwrites the first. I imagine it would be helpful to add multiple files at the same time for uploading of assignments. But that said you may have built this with scoping in mind and know that isn't something we need to do!

**Comment by stevenbister** [comment]
> Looks great mate. A couple of comments/questions in regards to the functionality.
> 
> * Looks like you can't remove an upload once you've added it? Think this could be quite important/helpful as it's easy to accidentally at the wrong file or something.

Yeah I was just mimic-ing the native file input behaviour, which just lets you re-upload over the top. Can have a look at implementing this behaviour though, is a good shout.

> * Is there a reason there is a pop up once it's been uploaded? Doesn't feel necessary to me given that you see the name of the file come up up in the download box anyway.

Yeah, it's just in storybook to demo that you can create actions based `onValueChange` -- it can be whatever

> * Doesn't look like you can upload multiple files at the same time - If I add a second it just overwrites the first. I imagine it would be helpful to add multiple files at the same time for uploading of assignments. But that said you may have built this with scoping in mind and know that isn't something we need to do!

You can, you just need to add the `multiple` prop. I'll add a story for this if it's not obvious

File: packages/osc-ui/src/components/FileInput/FileInput.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    });
+
+    const event = Object.defineProperty(dragDropEvent, 'dataTransfer', {
+        value: {
+            files: [file],
+        },
+    });
+
+    return event;
+};
```
Comment (hi-im-si) [inline packages/osc-ui/src/components/FileInput/FileInput.spec.tsx:97]: I am sure you have attempted to use ChatGPT or equiv to figure this issue out. I cam e across a few articles and put these through chagpt alongside your existing code and it came out with the following solution. I dont expect it to work but might be worth dropping in and testing - if not then easily discarded.

```suggestion
test('shows success icon when valid item is dropped', () => {
  // Mock the DataTransfer object
  class DataTransfer {
    constructor() {
      this.items = [];
    }

    setData(type, data) {
      this.items.push({ type, data });
    }

    getData(type) {
      const item = this.items.find((i) => i.type === type);
      return item ? item.data : null;
    }
  }

  const dataTransfer = new DataTransfer();

  // Create a file to simulate dropping
  const file = new File(['file'], 'test.png', {
    type: 'image/png',
  });

  // Simulate the drag event
  const dropZone = screen.getByTestId('file-input'); // Use your actual test ID
  const fileDropEvent = createEvent.drop(dropZone);

  Object.defineProperty(fileDropEvent, 'dataTransfer', {
    value: dataTransfer,
  });

  // Add the file to the dataTransfer
  dataTransfer.setData('Files', file);

  render(<FileInput id="file-input" />);

  fireEvent(dropZone, fileDropEvent);

  expect(screen.getByText('test.png')).toBeInTheDocument();
});
```

**Review by hi-im-si — CHANGES_REQUESTED** [review]

File: packages/osc-ui/src/components/FileInput/FileInput.spec.tsx (hunk 2 above)
Comment (stevenbister) [reply packages/osc-ui/src/components/FileInput/FileInput.spec.tsx:97]: Yeah, no dice I'm afraid :/

**Review by mattwendzina — COMMENTED** [review]
Looking good. Just added one more comment.

**Comment by stevenbister** [comment]
> Nice. I can see this story, but was initially confused as it still would only allow one. Turns out that multiple is still set to false in the story which is strange as you've set multiple to 'true' 🤔

Yeah Storybook it being weird at the moment with that 🤔 I think it might have something to do with the upgrade, presumably once we get around to updating the stories to the new format it'll sort itself out 🤞 

> It does work though. My only nitpick with it would be that I think it would be nice to be able to add multiple one at a time as well as being able to add multiple in one go. From what I can see it looks like it just overrides the old one if you add one at a time. Think there could be scenarios when someone might want to upload multiples but they aren't necessarily in the same folder or easy to add in one go.
> 
> Guess this is a nice to have though so if that's too much work to do then we could leave this off. Happy to approve as it if so.

Yeah unfortunately that's not really how the `input type=file` tag works :/ If you use the button it just hooks into that and I can't change that so I don't really want to have different behaviour for drag and drop as I can see that becoming a pain to maintain :/ Also, just generally I'd like to try and keep our components as close to the native html tags as possible

**Review by mattwendzina — APPROVED** [review]
LGTM

**Review by hi-im-si — APPROVED** [review]

**Pull request merged**